### PR TITLE
Replace Painless's Cast with casting strategies

### PR DESCRIPTION
**Pull request by nik9000**
Replaces the `Cast` object which had 7 parameters with casting
strategy objects that ought to be much easier to read. The trouble
with the `Cast` object is that it is fairly difficult to figure
out from reading the constructor invocation what the instructions
are emitted for the cast. The casting strategies are much easier
to read. Compare:
```
- return new Cast(LONG_TYPE, expected, true, LONG_TYPE, null, null, null);
+ return new Cast.Unbox(LONG_TYPE, new Cast.Numeric(LONG_TYPE, expected));
```

They both do the same thing but the bottom one can be read
"unbox into a `long` and then demote to the expected type".

Or compare:
```
- return new Cast(BOOLEAN_OBJ_TYPE, DEF_TYPE, explicit, null, null, BOOLEAN_TYPE, null);
+ return new Cast.Box(BOOLEAN_TYPE);
```

Both of them do the same thing but the bottom one reads
"box a boolean" while the top one is kind of an eye chart.

I reordered many of the `case` clauses so that they were
in a consistent order: primitives (smallest to largest),
`def`, `Object`, `Number`, boxed numbers (smallest to
largest). I tried to be consistent though I imagine I wasn't
always successful.

I also added a ton of test cases for casting. There are four
sorts of "regular" casts in painless, and I made sure to cover
them fairly extensively.

* "implicit non-internal": Generally only works when promoting
numbers or widening types. For example, assigning the
`int` `1` to `long`:
```
long n = 1;
```

* "implicit internal": Like "implicit, non-internal" but also
supports boxing. Generally used for return values. For
example, a function returning an `Integer` can accept
the `int` `1`:
```
Integer foo() {return 1}
```

* "explicit non-internal": Fine for promoting or demoting
numbers and widening or narrowing types. For example,
assigning the `long` `1` to an `int`:
```
long l = 1L;
int n = (int) l;
```

* "explicit internal": Fine for promoting or demoting
numbers, widening or narrowing types, and boxing.
Interestingly, you can only trigger these in a foreach
loop:
```
int[] array = new int[1]
array[0] = 1
for (Short s : array) {   // see the demotion and boxing?!
```

I labeled this a `bug` because it fixes a few issues that I found while testing the "explicit, internal" casts.

### CODE REVIEW

**Comment by nik9000** [comment]
Also: I've intentionally tried not to change any behavior if possible. If we think we should change behavior, specifically around "explicit internal", I think we can do it either here or in a followup.....

**Comment by jdconrad** [comment]
When we discussed the issue, we agreed that the case statement in AnalyzerCaster was very difficult to read based on the number of parameters that were going into a Cast object.  After giving this some thought, I believe this can be solved in a simpler way by having four static methods in Cast that act as factory methods -- "boxThenCast", "castThenBox," "unboxThenCast,", and "castThenUnbox."  Finally, a "castUsingMethod" can be added to resolve the method calls that can be made.  The explicit parameter can be replaced with a method parameter.  Right now, there are no cast chains that require more than a box/unbox and a singular cast (and I really hope there never will be), so having 4+ new classes containing what amounts to a custom linked list of casts seems like overkill for this.  

This change also leaks ASM into the analysis phase which (while there are probably a few unfortunate exceptions) doesn't have to happen here.  We tried to contain ASM to four places -- Definition, write methods in the nodes, possibly Def/DefBootstrap, and MethodWriter.  Analysis is pretty careful to only use Painless types and such, so that all ASM code remains nicely isolated. I think we can add a CastMethod Map or something along those lines in Definition for the DefImplicit/DefExplicit casts to contain any ASM outside of AnalyzerCaster.

The re-ordering of the case statements is fine with me.  I had them ordered based on internal and explicit casts primarily and then size of the numeric secondarily.

Finally, on style -- we have very different subjective coding styles as did Robert and Uwe which is really starting to show.  I noticed that you removed the newlines after if statements but left the break on the next line in the case statement which is inconsistent with the other files.  I'm happy to be flexible here since I know Elastic has no official style guide, but we should develop one for Painless so that the code looks consistent.

Tests -- huge ++, thanks for adding them.

File: modules/lang-painless/src/main/java/org/elasticsearch/painless/node/AExpression.java [hunk 1]
```diff
@@ -137,7 +137,7 @@ AExpression cast(Locals locals) {
                 EConstant econstant = new EConstant(location, constant);
                 econstant.analyze(locals);
 
-                if (!expected.equals(econstant.actual)) {
+                if (!expected.clazz.isAssignableFrom(econstant.actual.clazz)) {
```
Comment (jdconrad) [inline modules/lang-painless/src/main/java/org/elasticsearch/painless/node/AExpression.java]: Why did this change?

**Comment by nik9000** [comment]
> having 4+ new classes containing what amounts to a custom linked list of casts seems like overkill for this.

I've grown kind of attached to the side effects that come with the classes. Removing the if statements in MethodWriter feels like a win. I think debugging will be easier with the more descriptive `toString`s in the casting strategies and having fewer fields on the casting strategies than the old `Cast`. Is there any way you'd be ok with keeping them? If we simplified them or hid them behind factory methods?

> This change also leaks ASM into the analysis phase

Makes sense to me. I can totally see getting ASM out of analyzer caster. If we can keep the casting strategies, can they have ASM in them?


> Elastic has no official style guide, but we should develop one for Painless so that the code looks consistent

I'm happy to live with whatever you like. I have my own opinions but I don't hold them so tightly that I'd be willing to try to convince people of them.

The rest of Elasticsearch very rarely uses single line if statements instead of block if statements. At least in code written in the past few years. When it does they are almost always `break` or `return` and they are almost always on the same line. So I was brining `AnalyzerCaster` in line with what I've seen in the rest of Elasticsearch. But I'm really ok with whatever you think makes it more readable. If we want different and/or stricter rules than the rest of Elasticsearch we can do that though I'd have to think about enforcement.

As for the `switch`/`case` statements: Elasticsearch doesn't really have rules around those. If you want them formatted a certain way I'm happy to do that. Personally, I don't like indenting `case` clauses but I kept it because it started that way.

**Comment by jdconrad** [comment]
I understand the classes are important to you, so if you really want them, I won't block anything here.  I do think we are losing some things, though.  The first is stylistically we no longer follow the patterns ASM MethodWriter (and other ASM classes usually follow) by having case/if statements for most things that are type related.  The second is having the ASM for casting all located in one place where we would end up with ~30 lines of code in an if/else block versus spread across 4+ classes and nearly ~250 lines of code.  (I'm really not sure removing an if/else block here justifies the cost.)  Third, box/unbox requires two separate classes to be instantiated in a linked list style format.  And fourth, there was already a Cast class, so a toString could be added there pretty easily for debugging.

If the classes are kept, then it of course makes sense to have a write method there.  But please do *not* add any factory methods; they are completely unnecessary since multiple constructors now exist for each of the casts.  

I'm not that picky about style either as long as it's consistent.  I'll see if I can find an outline for a style guide (Java does have one, doesn't it?) that we can at least use.  Since Painless is likely to become it's own project I think it would be nice to enforce this outside of the main ES project for anyone else that ends up working on Painless with us.  It's especially nice to be able to point to something about style without being subjective :)

My requested changes:
* Please add something to the Definition to prevent ASM leaking into AnalyzerCaster.
* Not a change, but I commented above on a code change that I don't fully understand.  If you have time, I would appreciate an example or two for why that changed.

File: modules/lang-painless/src/main/java/org/elasticsearch/painless/node/AExpression.java (hunk 1 above)
Comment (nik9000) [inline modules/lang-painless/src/main/java/org/elasticsearch/painless/node/AExpression.java]: I've pushed 8f05769a6ca8cf5d566d54e7e74f8c05fd140332.

**Comment by rjernst** [comment]
Sorry for the late comment here; I am catching up on emails. I do not think we should add this complexity of all these new classes. What was there before was simple and straightforward. The debugging argument with `toString()` could be done on the current `Cast` class, and the ctor made private with static factory methods which achieve the equivalent of the classes/list being added here. Having a list of these just adds the possibility for us to mess it up.  Let's stick with simple until we find a compelling reason to add complexity.

**Comment by nik9000** [comment]
I disagree that what we have is simpler. Certainly shorter but harder to read and reason about. But I'm willing to give up on it because neither of you like it.

**Comment by rjernst** [comment]
@nik9000 Being shorter is exactly what makes it easier to read and reason about. It is contained, and you don't have to think about what a particular instance of some base class might actually be.

I do think there was some good stuff in this PR like having better toString(), and some tests. Do you maybe want to go with the factory methods, fold in toString() into the existing class with cases based on all the members, and keep some tests?

**Comment by nik9000** [comment]
> @nik9000 Being shorter is exactly what makes it easier to read and reason about. It is contained, and you don't have to think about what a particular instance of some base class might actually be.

But you can't look at an instance of Cast and figure out what it is going to try to do. And you can build instances of Cast that function in unexpected ways. The factory method approach will help with the second one but not the first.



> I do think there was some good stuff in this PR like having better toString(), and some tests. Do you maybe want to go with the factory methods, fold in toString() into the existing class with cases based on all the members, and keep some tests?

I can see about the tests and the factory methods. I don't think the toString is a good idea without the classes because it'd be a reimplemention of the big if statement tree for casting.

**Pull request closed without merging**